lib/Controller/Action.php

```diff
@@ -248,17 +256,28 @@ public function addForm(Request $request, Response $response, string $source, in
 
         // Make sure the Layout is checked out to begin with
         if (!$layout->isEditable()) {
-            throw new InvalidArgumentException(__('Layout is not checked out'), 'statusId');
+            throw new InvalidArgumentException(__('Layout is not checked out'), 'publishedStatusId');
         }
 
         $layout->load();
 
+        // all widgets
+        $widgets = $layout->getAllWidgets();
+
+        foreach ($widgets as $widget) {
+            $module = $this->moduleFactory->createWithWidget($widget);
+            // if we don't have a name set in the Widget
+            $widget->name = $module->getName();
+        }
+
         $this->getState()->template = 'action-form-add';
         $this->getState()->setData([
             'sourceObject' => $sourceObject,
             'source' => $source,
             'id' => $id,
-            'regions' => $layout->regions
+            'regions' => $layout->regions,
+            'widgets' => $widgets,
+            'layoutCodes' => $this->layoutFactory->getLayoutCodes()
```

Comment:
This could potentially be 1000s of Layouts?
Would it be better for this to be a dynamic select2 with paging?